cmd/snap/cmd_debug_state.go

```diff
@@ -293,6 +301,165 @@ func (c *cmdDebugState) showIsSeeded(st *state.State) error {
 	return nil
 }
 
+// connState is the connection state stored by InterfaceManager
+type connState struct {
+	Auto             bool                   `json:"auto,omitempty" yaml:"auto"`
+	ByGadget         bool                   `json:"by-gadget,omitempty" yaml:"by-gadget"`
+	Interface        string                 `json:"interface,omitempty" yaml:"interface"`
+	Undesired        bool                   `json:"undesired,omitempty" yaml:"undesired"`
+	StaticPlugAttrs  map[string]interface{} `json:"plug-static,omitempty" yaml:"plug-static,omitempty"`
+	DynamicPlugAttrs map[string]interface{} `json:"plug-dynamic,omitempty" yaml:"plug-dynamic,omitempty"`
+	StaticSlotAttrs  map[string]interface{} `json:"slot-static,omitempty" yaml:"slot-static,omitempty"`
+	DynamicSlotAttrs map[string]interface{} `json:"slot-dynamic,omitempty" yaml:"slot-dynamic,omitempty"`
+	// TODO: hotplug
+}
+
+type connectionInfo struct {
+	PlugSnap string
+	PlugName string
+	SlotSnap string
+	SlotName string
+
+	connState
+}
+
+type byPlug []*connectionInfo
+
+func (c byPlug) Len() int      { return len(c) }
+func (c byPlug) Swap(i, j int) { c[i], c[j] = c[j], c[i] }
+func (c byPlug) Less(i, j int) bool {
+	a, b := c[i], c[j]
+	return a.PlugSnap < b.PlugSnap || (a.PlugSnap == b.PlugSnap && a.PlugName < b.PlugName)
+}
+
+func (c *cmdDebugState) showConnectionDetails(st *state.State, connArg string) error {
+	st.Lock()
+	defer st.Unlock()
+
+	p := strings.FieldsFunc(connArg, func(r rune) bool {
+		return r == ' ' || r == ','
+	})
+
+	var plugMatch, slotMatch SnapAndName
+	if err := plugMatch.UnmarshalFlag(p[0]); err != nil {
+		return err
+	}
+
+	if len(p) > 1 {
+		if err := slotMatch.UnmarshalFlag(p[1]); err != nil {
+			return err
+		}
+	}
+
+	var conns map[string]*connState
+	if err := st.Get("conns", &conns); err != nil && !errors.Is(err, state.ErrNoState) {
+		return err
+	}
+
+	// sort by connection ID
+	connIDs := make([]string, 0, len(conns))
+	for connID := range conns {
+		connIDs = append(connIDs, connID)
+	}
+	sort.Strings(connIDs)
+
+	for _, connID := range connIDs {
+		connRef, err := interfaces.ParseConnRef(connID)
+		if err != nil {
+			return err
+		}
+
+		refMatch := func(x SnapAndName, y interface{ String() string }) bool {
+			parts := strings.Split(y.String(), ":")
+			return len(parts) == 2 && x.Snap == parts[0] && x.Name == parts[1]
+		}
+		plug, slot := connRef.PlugRef, connRef.SlotRef
+
+		if slotMatch.Name != "" && slotMatch.Snap != "" && plugMatch.Snap != "" && plugMatch.Name != "" {
```

Comment:
(nitpick^3) I would use a `switch` command here, mostly because it's slightly more compact and put the same order everywhere:
```
	switch {
		case slotMatch.Snap != "" && slotMatch.Name != "" && plugMatch.Snap != "" && plugMatch.Name != "":
			// command invoked with 'snap:plug,snap:slot'
			if !refMatch(plugMatch, plug) || !refMatch(slotMatch, slot) {
				// should match the connection exactly
				continue
			}
		case slotMatch.Snap == "" && slotMatch.Name == "" && plugMatch.Snap != "" && plugMatch.Name != "":
			// command invoked with 'snap:plug-or-slot'
			if !refMatch(plugMatch, plug) && !refMatch(plugMatch, slot) {
				// should match either the connection's slot or plug
				continue
			}
		case slotMatch.Snap == "" && slotMatch.Name == "" && plugMatch.Snap != "" && plugMatch.Name == "":
			// command invoked with 'snap' only
			if plugMatch.Snap != slot.Snap && plugMatch.Snap != plug.Snap {
				// should match one of the snap names
				continue
			}
		default:
			return fmt.Errorf("invalid command with connection args: %s", connArg)
		}
```
(with that one can also see that `plugMatch.Snap == ""` is always an error so we could check that earlier and omit it from this switch/case but it's probably fine to just keep it here)

Reply:
Yeah, spelling them out fully is nice, less state in the brain is always (for me at least) better :)